client/components/site-verticals-suggestion-search/index.jsx

```diff
@@ -26,55 +26,126 @@ import './style.scss';
-			fetch_algo: '/verticals',
-			action: 'site_vertical_selected',
-		};
+	componentDidUpdate( prevProps ) {
+		// The suggestion list should only be updated when the vertical suggestions results change.
+		// Note: it's intentional to use object reference comparison here.
+		// Since `verticals` props is connected from a redux state here, if the two references are identical,
+		// we can safely say that the two content are identical, thanks to the immutability invariant of redux.
+		if ( prevProps.verticals !== this.props.verticals ) {
+			// It's safe here to call setState() because we prevent the indefinite loop by the wrapping condition.
+			// See the official doc here: https://reactjs.org/docs/react-component.html#componentdidupdate
+			this.setSearchResults( this.props.verticals );
+		}
 	}
 
-	// When a user is keying through the results,
-	// only update the vertical when they select a result.
+	/**
+	 * Show the popular topics component when the search field is empty and
+	 * `props.showPopular` tell us to.
+	 *
+	 * @returns {Bool} Whether we should display the popular topics component.
+	 */
+	shouldShowPopularTopics = () => ! this.state.inputValue && this.props.showPopular;
+
+	/**
+	 * Checks for the existence of vertical results. If there are none we assume
+	 *
+	 * @returns {Bool} Whether we should display the popular topics component.
+	 */
+	isVerticalSearchPending = () => this.state.inputValue && 0 === this.props.verticals.length;
+
+	/**
+	 * Sets `state.results` with incoming vertical results, retaining previous non-user vertical search results
+	 * if the incoming vertical results contain only user-defined results.
+	 *
+	 * This function could better be performed in the backend eventually.
+	 *
+	 * @param {Array} results Incoming vertical results
+	 */
+	setSearchResults = results => {
+		if ( results && results.length ) {
+			const { candidateVerticals, inputValue, isSuggestionSelected } = this.state;
+			// If the user is typing
+			// and the only result is a user input (non-vertical),
+			// and we have some previous results
+			// then concat that with the previous results and remove the last user input
+			if (
+				! isSuggestionSelected &&
+				! find( results, item => ! item.isUserInputVertical ) &&
+				1 < candidateVerticals.length
+			) {
+				results = candidateVerticals.filter( item => ! item.isUserInputVertical ).concat( results );
+			}
+
+			this.setState( { candidateVerticals: results }, () =>
+				this.updateVerticalData( this.searchForVerticalMatches( inputValue ), inputValue )
+			);
+		}
+	};
+
+	getNewRailcar = () => ( {
+		id: `${ uuid().replace( /-/g, '' ) }-site-vertical-suggestion`,
+		fetch_algo: '/verticals',
+		action: 'site_vertical_selected',
+	} );
+
+	/**
+	 * Searches the API results for a direct match on the user search query.
+	 *
+	 * @param {String} value Search query array
+	 * @returns {Object?} An object from the vertical results array
```

Comment:
Docs nit: unless I'm missing something, this function will return `undefined` if there are no matches, and not `null`. We should reflect that in the docs (removing the `?`).

In case I'm right about this, let's make sure any usages of `searchForVerticalMatches` work well with the `undefined` case.